### PR TITLE
Keep track of allocation and release memory (faster than by calling mallinfo and more accurate than calling footprint)

### PR DESCRIPTION
**Pull request by palainp**
This PR adds a counter to memory usage (simply malloc calls add the size value, and free calls substract the size value). The tricky part is that dmalloc can use `dispose_chunk` to do something similar to free and we have to substract memory used there too.

This doesn't change the API and just adds an exported function `malloc_memory_usage` (I'm not sure that's the best name though).

### CODE REVIEW

**Comment by hannesm** [comment]
Thanks for this PR. I was wondering whether the subtraction could be done at a single place, `dispose_chunk`, since it looks like with this PR it happens just before every call site thereof?

File: nolibc/dlmalloc.i [hunk 1]
```diff
@@ -5121,8 +5136,12 @@ static size_t internal_bulk_free(mstate m, void* array[], size_t nelem) {
             set_inuse(m, p, newsize);
             *b = chunk2mem(p);
           }
-          else
+          else {
```
Comment (hannesm) [inline nolibc/dlmalloc.i:5139]: in this function, `internal_bulk_free`, I don't quite understand why in the above consequence ("try to merge with next chunk") that `internal_memory_usage` is not updated?

Comment (palainp) [reply nolibc/dlmalloc.i:5139]: The `set_inuse(m, p, newsize);` marks the bloc `p` to be sized as `newsize`(=current size+next chunk size) and `*b = chunk2mem(p);` will prepare the next loop iteration (`b` is a pointer to the next cell in the array to be freed). We will substract the chunk size only when the next chunk cannot be merged to the current one.

Anyway, by moving the `internal_memory_usage` change into `dispose_chunk` this part of the patch will no longer appear :)

**Comment by palainp** [comment]
> Thanks for this PR. I was wondering whether the subtraction could be done at a single place, `dispose_chunk`, since it looks like with this PR it happens just before every call site thereof?

You're totally right about that, this comes from my tests to find out why the value of `internal_memory_usage` can goes down the real usage and unfortunately I changed two things at once:
* removing static and adding volatile qualifier for `internal_memory_usage`
* the extraction of `internal_memory_usage -= ...` from `dispose_chunk`
But it appears that I now have the substraction before every `dispose_chunk` calls. I'll test out with a single place modification at the begining of `dispose_chunk`.

**Comment by palainp** [comment]
After testing (intensive network usage + 32M unikernel which is barely enough to try to have the worst possible situation) I'm not able to reproduce any difference between the value returned by `mallinfo` and the `internal_memory_usage` value. It's better to declare it as static and with code factorization it is easier to read.

**Comment by palainp** [comment]
I think I need some advice here, looking back to the patch, with regard to the following paragraph 
https://github.com/mirage/ocaml-solo5/blob/80ff24ef8f678e2b2f7c6c8dc45e59caf739a0cb/nolibc/dlmalloc.i#L2990
it may be better to move the `internal_memory_usage` inside the global state:
https://github.com/mirage/ocaml-solo5/blob/80ff24ef8f678e2b2f7c6c8dc45e59caf739a0cb/nolibc/dlmalloc.i#L2587

**Comment by palainp** [comment]
I hope this notice comment will be clear and explanatory enough.

As a note, could it be possible to co-author @winux138 for [e9c5343](https://github.com/mirage/ocaml-solo5/pull/120/commits/e9c53438f7a26a46a4d8354f4a9b3cf42f59b2e1) and [80ff24e](https://github.com/mirage/ocaml-solo5/pull/120/commits/80ff24ef8f678e2b2f7c6c8dc45e59caf739a0cb) who worked on that too ? I'm not sure I have done that correctly.

**Comment by hannesm** [comment]
This looks fine to me. I also looked into the difference of assembly output between main branch and these changes, and there is not much difference (some subs etc., but not any more function calls etc.).

> As a note, could it be possible to co-author @winux138 for [e9c5343](https://github.com/mirage/ocaml-solo5/pull/120/commits/e9c53438f7a26a46a4d8354f4a9b3cf42f59b2e1) and [80ff24e](https://github.com/mirage/ocaml-solo5/pull/120/commits/80ff24ef8f678e2b2f7c6c8dc45e59caf739a0cb) who worked on that too ? I'm not sure I have done that correctly.

You did it at least partially correct, what works is the line in a commit message
`Co-authored-by: NAME <email-address>`

Now, you only put "@winux138" there which relies on GitHub to translate the user ID to a link -- if you put the eMail address there instead, it should work (and be independent of GitHub as a plaform).

Would you also mind to squash all commits into a single one?

File: nolibc/dlmalloc.i [hunk 2]
```diff
@@ -964,6 +975,13 @@ DLMALLOC_EXPORT void* dlvalloc(size_t);
 */
 DLMALLOC_EXPORT int dlmallopt(int, int);
 
+/*
+  var init for malloc_memory_usage
+  We need to declare it as extern and reserve memory for that in the kernel
+  as for sbrk()
```
Comment (hannesm) [inline nolibc/dlmalloc.i:981]: Could you describe here (a) what does it return (presumably "Returns the number of bytes obtained from the system.") and (b) what's the difference to `mallinfo` and `malloc_footprint`? I'd guess "Unlike mallinfo, this functions returns a precomputed value that is updated on each call to malloc and free. Unlike malloc_footprint ???".

**Pull request merged**